### PR TITLE
fix: handle batched requests in sagemaker embedding endpoint

### PR DESCRIPTION
**Pull request by evandiewald**
Related to: #3211 

`SagemakerEndpointEmbeddings.embed_documents` batches documents and invokes the endpoint once per batch, meaning that the output of `_embedding_func` should be a *list of lists*, or one embedding vector per document. The current behavior will only return one vector *per batch*, rather than one vector per document.

The issue is that if you follow the [linked documentation](https://www.philschmid.de/custom-inference-huggingface-sagemaker) for setting up an embedding endpoint, the endpoint will only return the results for one input at a time. 

From the custom `inference.py` file in the tutorial, `predict_fn` only returns the embedding vector for the first input:

```
# inference.py
...
def predict_fn():
    ...
    return {"vectors": sentence_embeddings[0].tolist()}
```

So to handle batched requests (i.e. lists of text snippets), you simply have to change this return line to:

```
# return list of lists
return {"vectors": sentence_embeddings.tolist()}
```

The other change you have to make is to the `ContentHandler`, as `transform_input` should accept a list of strings as used in `_embedding_func`. I updated the documentation to show how this works. 

The alternative is to get rid of the request batching and just make one `invoke_endpoint` call per document, but I agree that batching is the better way to go as long as your endpoint can support it. 

Hope this makes sense! Love the project and happy to adjust based on design preferences.

### CODE REVIEW

**Comment by evandiewald** [comment]
#2631 also seems like the same bug

File: langchain/embeddings/sagemaker_endpoint.py [hunk 1]
```diff
@@ -71,17 +71,17 @@ class SagemakerEndpointEmbeddings(BaseModel, Embeddings):
...
+
         from langchain.llms.sagemaker_endpoint import ContentHandlerBase
 
         class ContentHandler(ContentHandlerBase):
                 content_type = "application/json"
                 accepts = "application/json"
 
-                def transform_input(self, prompt: str, model_kwargs: Dict) -> bytes:
-                    input_str = json.dumps({prompt: prompt, **model_kwargs})
+                def transform_input(self, prompts: List[str], model_kwargs: Dict) -> bytes:
```
Comment (vowelparrot) [inline langchain/embeddings/sagemaker_endpoint.py:81]: The abstract class was updated to take `Union[str, List[str]]`, so we'll need to reflect that here. Thanks for the change!

**Review by vowelparrot — COMMENTED** [review]
Thanks for the PR! I'm going to make some changes to the notebook and classes to build on this fix. Have you tested this on any endpoints?

**Comment by evandiewald** [comment]
> Thanks for the PR! I'm going to make some changes to the notebook and classes to build on this fix. Have you tested this on any endpoints?

Happy to contribute! I have tested it on an endpoint. Something like this works as expected:

```
class ContentHandler(ContentHandlerBase):
    content_type = "application/json"
    accepts = "application/json"

    def transform_input(self, prompts: list[str], model_kwargs: Dict) -> bytes:
        input_str = json.dumps({"inputs": prompts, **model_kwargs})
        return input_str.encode('utf-8')

    def transform_output(self, output: bytes) -> list:
        response_json = json.loads(output.read().decode("utf-8"))
        return response_json["vectors"]


content_handler = ContentHandler()

embeddings = SagemakerEndpointEmbeddings(
    endpoint_name="EmbeddingEndpointHfEndpoint....",
    region_name="us-east-1",
    content_handler=content_handler
)

loader = TextLoader("example.txt")
text_splitter = CharacterTextSplitter(chunk_size=500, chunk_overlap=0, separator=" ")
documents = loader.load()

split_docs = text_splitter.split_documents(documents=documents)
texts = [d.page_content for d in split_docs]

emb = embeddings.embed_documents(texts)
assert len(texts) == len(emb)

embedded_query = embeddings.embed_query("foo")
print(embedded_query) # [-0.01905810832977295, 0.022155532613396645, -0.048968423157930374,...]
```

**Comment by vowelparrot** [comment]
I've made some changes to make it a bit more clear what to expect (`Any` typing was confusing) - what are your thoughts on the latest commit?

**Pull request merged**